File: android/app/src/main/kotlin/com/example/perfect_corp_homework/service/DownloadServiceImpl.kt
package com.example.perfect_corp_homework.service

import com.example.perfect_corp_homework.MainActivity
import com.example.perfect_corp_homework.api.ServiceResult
import com.example.perfect_corp_homework.model.DownloadEntity
import com.example.perfect_corp_homework.model.DownloadStatus
import com.example.perfect_corp_homework.repository.DownloadRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import java.util.function.Consumer
import kotlin.coroutines.CoroutineContext

class DownloadServiceImpl constructor(private val downloadRepository: DownloadRepository): DownloadService {

    private val log = LoggerFactory.getLogger("DownloadServiceImpl")

    private var id2DownloadEntity = mutableMapOf<String, DownloadEntity>()

    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main

    override suspend fun createDownload(urlString: String) : ServiceResult<Int>{

        try {
            val downloadEntity = async(Dispatchers.Default) {
                downloadRepository.fetchImageMetadata(
                    urlString = urlString,)
            }.await()

            id2DownloadEntity+=(downloadEntity.downloadID to downloadEntity)
            withContext(Dispatchers.Main) {
                MainActivity.sendProgressEvent(id2DownloadEntity.values.toList())
            }

            startDownload(downloadEntity)

            return ServiceResult<Int>(data = calculateOngoingDownload())
        } catch (e: Exception) {
            return ServiceResult<Int>(error = e)
        }
    }


    val updateProgress = { pair: Pair<String, Int> ->
        id2DownloadEntity[pair.first]!!.updateProgress(pair.second)
        MainActivity.sendProgressEvent(id2DownloadEntity.values.toList())
    }

    private fun calculateOngoingDownload(): Int {
        return id2DownloadEntity.values.filter { it ->  it.status == DownloadStatus.ongoing}.size
    }


    override suspend fun startDownload(downloadEntity: DownloadEntity): Unit {
        try {
            launch(Dispatchers.Default) {
                withContext(Dispatchers.Main) {
                    downloadEntity.status = DownloadStatus.ongoing
                    MainActivity.sendProgressEvent(id2DownloadEntity.values.toList())
                }
                downloadRepository.fetchImageWithUrl(downloadEntity = downloadEntity, updateProgress)
            }
        } catch (e: Exception) {
            throw e
        }
    }


    private fun ongoingDownloadCount(): Int {
        return id2DownloadEntity.values.filter { it.status == DownloadStatus.ongoing }.size
    }

    override suspend fun pauseDownload(downloadID: String): ServiceResult<Int> {
        val target = id2DownloadEntity[downloadID]
        target!!.pauseDownload()
        withContext(Dispatchers.Main){
            MainActivity.sendProgressEvent(id2DownloadEntity.values.toList())
        }
        return ServiceResult(ongoingDownloadCount())
    }

    override suspend fun resumeDownload(downloadID: String): ServiceResult<Int> {
        val target = id2DownloadEntity[downloadID]
        target!!.resumeDownload()
        withContext(Dispatchers.Main){
            MainActivity.sendProgressEvent(id2DownloadEntity.values.toList())
        }
        return ServiceResult(ongoingDownloadCount())
    }

    override suspend fun cancelDownload(downloadID: String): ServiceResult<Int> {
        val target = id2DownloadEntity[downloadID]
        target!!.cancelDownload()
        withContext(Dispatchers.Main){
            MainActivity.sendProgressEvent(id2DownloadEntity.values.toList())
        }
        return ServiceResult(ongoingDownloadCount())
    }

    override suspend fun manualPauseDownload(downloadID: String): ServiceResult<Int> {
        val target = id2DownloadEntity[downloadID]
        target!!.manualPauseDownload()
        withContext(Dispatchers.Main){
            MainActivity.sendProgressEvent(id2DownloadEntity.values.toList())
        }
        return ServiceResult(ongoingDownloadCount())
    }
}

File: plugin/download/ios/Classes/service/DownloadServiceImpl.swift
//
//  DownloadServiceImpl.swift
//  Runner
//
//  Created by SamLS Chen on 2024/8/19.
//

import Foundation

class DownloadServiceImpl: DownloadService {
    
    
    private var id2DownloadEntity: [String: DownloadEntity] = [:]
    private let downloadRepository: DownloadRepository
    
    init(downloadRepository: DownloadRepository) {
        self.downloadRepository = downloadRepository
    }
    
    func createDownload(urlString: String, isConcurrent: Bool) async -> ServiceResult<Int> {
        do {
            let downloadEntity = try await downloadRepository.configureDownload(from: urlString, isConcurrent: isConcurrent)
            id2DownloadEntity[downloadEntity.downloadID] = downloadEntity
            
            
            await updateProgress()
            await startDownload(downloadEntity: downloadEntity)
            await updateProgress()
            
            return ServiceResult<Int>(data: _calculateOngoingDownload())
        } catch {
            if(error is AppError) {
                return ServiceResult<Int>(error: error as! AppError)
            }
            return ServiceResult<Int>(error: AppError.UnknownError(error.localizedDescription))
        }
        
    }
    
    func updateProgress() async {
        do {
            try await MainActor.run {
                DownloadPlugin.progressEventSink!(try JSONSerialization.jsonObject(with: try JSONEncoder().encode(Array(id2DownloadEntity.values.sorted(by: { d1, d2 in
                    d1.createAt<d2.createAt
                })))))
            }
            
        } catch {
            
        }
    }
    
    private func _calculateOngoingDownload() -> Int {
        return id2DownloadEntity.values.filter{ downloadEntity in
            downloadEntity.status==DownloadStatus.ongoing
        }.count
    }
    
    func startDownload(downloadEntity: DownloadEntity) async {
        do {
            try downloadRepository.fetchImage(baseOn: downloadEntity, updateProgress: updateProgress)
            downloadEntity.status = DownloadStatus.ongoing
        } catch {
            
        }
        
        
    }
    
    func resumeDownload(downloadID: String) async -> ServiceResult<Int> {
        let target = id2DownloadEntity[downloadID]
        target!.resumeDownload()
        await updateProgress()
        return ServiceResult<Int>(data: _calculateOngoingDownload())
    }
    
    func cancelDownload(downloadID: String) async -> ServiceResult<Int> {
        let target = id2DownloadEntity[downloadID]
        do{
            try target!.cancelDownload()
        } catch {
            return ServiceResult(error: AppError.UnknownError(error.localizedDescription))
        }
        
        await updateProgress()
        return ServiceResult<Int>(data: _calculateOngoingDownload())
        
    }
    
    func pauseDownload(downloadID: String) async -> ServiceResult<Int> {
        let target = id2DownloadEntity[downloadID]
        target!.pauseDownload()
        await updateProgress()
        return ServiceResult<Int>(data: _calculateOngoingDownload())
    }
    
    func manualPauseDownload(downloadID: String) async -> ServiceResult<Int> {
        let target = id2DownloadEntity[downloadID]
        target!.manualPausedDownload()
        await updateProgress()
        return ServiceResult<Int>(data: _calculateOngoingDownload())
    }
    
    
    
    
}
